File: web/src/api/endpoints/prompt.ts
import { apiFetch } from '@/api/client'
import type { MenuSelections } from '@/constants/menu_options'

/**
 * `POST /v1/prompt/preview` — see `planning/backend/api-shape.md` §5.6.
 * Pure read: combines fixed platform constraints, menu fragments, and the
 * LLM-reconciled English note into the final model prompt without firing
 * a generation. Backend caches the reconciler call in Redis (24h TTL) so
 * reopens of the same input are cheap.
 *
 * Request shape is a discriminated union by `mode` (T-035 backend contract).
 * Hand-coded here so frontend Wave A can land in parallel with backend;
 * once T-035 merges, run `pnpm typegen` and replace the union with the
 * generated types from `src/api/generated/openapi-types.ts`.
 */
export type PromptPreviewMode = 'create_base' | 'create_alias' | 'create_motion'

/** Mask reference (T-035 `MaskInput`). Mask payload is uploaded out-of-band
 *  and referenced by id; preview only needs to know one is present. */
export interface MaskInput {
  mask_id: string
}

export type AliasInputMode = 'text' | 'image' | 'inpaint' | 'mixed'

export type MotionParentType = 'base' | 'alias'

/** `mode='create_base'`. Used by the Creation Session page (T-022 / T-023).
 *  `base_checkpoint_id` is set when the user picks a checkpoint as the
 *  remix anchor — the worker reconciles with `has_reference_image=True`
 *  sourced from that checkpoint, and T-035 mirrors the signal in preview
 *  so the modal renders a faithful prompt (closes STATUS.md S2-5). */
export interface PromptPreviewBaseRequest {
  mode: 'create_base'
  menu_selections?: MenuSelections | null
  freeform_note?: string | null
  reference_image_ids?: string[] | null
  base_checkpoint_id?: string | null
}

/** `mode='create_alias'`. Used by the Alias edit page (T-036). */
export interface PromptPreviewAliasRequest {
  mode: 'create_alias'
  character_id: string
  input_mode: AliasInputMode
  menu_selections?: MenuSelections | null
  freeform_note?: string | null
  reference_image_ids?: string[] | null
  mask?: MaskInput | null
}

/** `mode='create_motion'`. Used by the Custom Motion modal (T-039); preset
 *  motions also flow through here so the modal can show the platform
 *  template. `motion_type` is `'preset_*'` or `'custom'`. */
export interface PromptPreviewMotionRequest {
  mode: 'create_motion'
  parent_type: MotionParentType
  parent_id: string
  motion_type: string
  description?: string | null
}

export type PromptPreviewRequest =
  | PromptPreviewBaseRequest
  | PromptPreviewAliasRequest
  | PromptPreviewMotionRequest

/** Mode-specific blocks (T-035 §Response schema) come back as optional
 *  fields rather than a separate response per mode — the consumer already
 *  knows which mode it asked for and narrows accordingly. */
export interface DerivedFromBase {
  base_id: string
  base_image_url: string
}

export interface MotionParent {
  type: MotionParentType
  id: string
  image_url: string
}

export type MotionTemplateUsed = 'custom_reconciled' | (string & {})

export interface PromptPreviewResponse {
  platform_constraints: string
  menu_fragments: string[]
  reconciled_note_en: string
  final_prompt: string
  /** Present when `mode='create_alias'`. */
  derived_from?: DerivedFromBase
  /** Present when `mode='create_motion'`. */
  parent?: MotionParent
  /** Present when `mode='create_motion'`. `'preset_*'` indicates the
   *  preset template was used as-is (no reconciler call). */
  motion_template_used?: MotionTemplateUsed
}

export function previewPrompt(body: PromptPreviewRequest): Promise<PromptPreviewResponse> {
  return apiFetch<PromptPreviewResponse>('/v1/prompt/preview', {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

File: api/app/schemas/prompt.py
"""DTOs for `POST /v1/prompt/preview` (T-019 + T-035).

Wire surface is a discriminated union by `mode` so the OpenAPI schema —
and any generated TS / Python client — narrows on the same field the
route uses to dispatch:

  - `create_base`  : the original Sprint-2 surface (T-019). Now also
    accepts `base_checkpoint_id` so a remix preview can faithfully reflect
    the worker-side image2image + has_reference_image=True path
    (closes STATUS.md backlog S2-5).
  - `create_alias` : Sprint-3 alias mode (text / image / inpaint / mixed).
    Carries `character_id` so the service can derive the base image
    surface, and an opt-in `mask: MaskInput` for inpaint/mixed.
  - `create_motion`: Sprint-3 motion mode (preset_* + custom). Carries
    `parent_type` + `parent_id` for the base/alias the motion will be
    attached to.

`MaskInput` is its own model so the upload-then-reference contract from
T-031 (`POST .../aliases/masks` returns `mask_id`, body uses
`{ mask: { mask_id } }`) shows up in OpenAPI rather than being hidden
behind a free-form `dict`. Closes STATUS.md backlog S3-1.
"""

from __future__ import annotations

import uuid
from typing import Annotated, Literal

from pydantic import BaseModel, Field

WirePromptMode = Literal["create_base", "create_alias", "create_motion"]

AliasInputMode = Literal["text", "image", "inpaint", "mixed"]

MotionParentType = Literal["base", "alias"]

MotionType = Literal[
    "preset_wave",
    "preset_nod",
    "preset_gesture",
    "preset_happy",
    "preset_idle",
    "custom",
]

# `motion_template_used` widens MotionType: preset_* values pass through
# verbatim (the worker reads the preset prompt from a static template),
# while `custom` is reported as `custom_reconciled` so callers can tell
# "user note went through the LLM" apart from "preset template applied".
MotionTemplateUsed = Literal[
    "preset_wave",
    "preset_nod",
    "preset_gesture",
    "preset_happy",
    "preset_idle",
    "custom_reconciled",
]


class MaskInput(BaseModel):
    """Reference to a mask uploaded via the alias-mask upload endpoint
    (T-031: `POST /v1/characters/{id}/aliases/masks` → returns
    `mask_id`).

    The wire field is `{ mask_id: UUID }` — only the identifier travels
    in the alias-create / preview body. The actual mask bytes stay in
    the storage backend until the alias worker reads them.
    """

    mask_id: uuid.UUID


class CreateBasePreviewRequest(BaseModel):
    mode: Literal["create_base"]
    # `dict[str, str]` (not `dict[str, Any]`) so the wire rejects e.g.
    # `{"age": 25}` upfront. The reconciler tolerates non-strings via a
    # `str(option)` fallback in `resolve_menu_fragments`, but the cache
    # key serialises `25` and `"25"` to different JSON, so a sloppy
    # frontend would fragment the cache for the same logical selection.
    menu_selections: dict[str, str] | None = None
    freeform_note: str | None = None
    reference_image_ids: list[uuid.UUID] | None = None
    # `base_checkpoint_id` is the source checkpoint a remix preview is
    # branching from. Its presence flips the worker into image2image
    # with `has_reference_image=True`; preview now mirrors that signal
    # so the modal renders the same prompt the worker would build.
    # Closes STATUS.md S2-5.
    base_checkpoint_id: uuid.UUID | None = None


class CreateAliasPreviewRequest(BaseModel):
    mode: Literal["create_alias"]
    character_id: uuid.UUID
    input_mode: AliasInputMode
    freeform_note: str | None = None
    reference_image_ids: list[uuid.UUID] | None = None
    mask: MaskInput | None = None


class CreateMotionPreviewRequest(BaseModel):
    mode: Literal["create_motion"]
    parent_type: MotionParentType
    parent_id: uuid.UUID
    motion_type: MotionType
    description: str | None = None


PromptPreviewRequest = Annotated[
    CreateBasePreviewRequest | CreateAliasPreviewRequest | CreateMotionPreviewRequest,
    Field(discriminator="mode"),
]


class DerivedFromInfo(BaseModel):
    """Alias-mode response surface — the base the alias will be derived
    from. `base_image_url` is a short-lived signed URL the modal uses
    as `<img src>`."""

    base_id: uuid.UUID
    base_image_url: str


class MotionParentInfo(BaseModel):
    """Motion-mode response surface — the base or alias the motion will
    animate. `image_url` is a signed URL of the parent's still frame."""

    type: MotionParentType
    id: uuid.UUID
    image_url: str


class PromptPreviewResponse(BaseModel):
    platform_constraints: str
    reconciled_note_en: str
    menu_fragments: list[str]
    final_prompt: str
    # Per-mode optional fields. Exactly one of these blocks is populated
    # per response (matching the request mode); the others stay None so
    # OpenAPI can keep a single response shape rather than a second
    # discriminated union purely for narrowing.
    derived_from: DerivedFromInfo | None = None
    parent: MotionParentInfo | None = None
    motion_template_used: MotionTemplateUsed | None = None
